// @flowusx/flowus-client/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/client.ts
var _flowusshared = require('@flowusx/flowus-shared');
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
var FlowUsClient = class {
  constructor(config) {
    this._baseUrl = (config == null ? void 0 : config.baseUrl) || "https://flowus.cn/api";
  }
  /**
   * 获取页面所有Block
   * @param pageId
   */
  getPageBlocks(pageId) {
    return __async(this, null, function* () {
      const pageBlocks = yield this._fetch(`docs/${pageId}`, { method: "GET" });
      pageBlocks.blocks = yield this.getOssUrl(pageBlocks.blocks);
      return pageBlocks;
    });
  }
  /**
   * 获取数据表格文档列表
   * @param pageId
   */
  getDataTableData(pageId) {
    return __async(this, null, function* () {
      const pageBlocks = yield this._fetch(`docs/${pageId}`, { method: "GET" });
      const blocksKeys = Object.keys(pageBlocks.blocks);
      const firstKey = blocksKeys[0];
      const firstValue = pageBlocks.blocks[firstKey];
      if (firstValue.type === _flowusshared.BlockType.Data_Table) {
        return pageBlocks;
      } else if (firstValue.type === _flowusshared.BlockType.Doc) {
        _flowusshared.out.err("\u7C7B\u578B\u9519\u8BEF", "\u8BF7\u4F7F\u7528 getPageBlocks \u83B7\u53D6\u9875\u9762\u6570\u636E");
      } else {
        _flowusshared.out.err("\u7C7B\u578B\u9519\u8BEF", "\u6682\u4E0D\u652F\u6301\u7684\u6587\u6863\u7C7B\u578B");
      }
      _process2.default.exit(-1);
    });
  }
  _fetch(endpoint, reqOpts) {
    return __async(this, null, function* () {
      var _a, _b;
      const url = `${this._baseUrl}/${endpoint}`;
      const res = yield _flowusshared.request.call(void 0, url, reqOpts);
      if (res.data.code !== 200) {
        if ((_b = (_a = res.data) == null ? void 0 : _a.msg) == null ? void 0 : _b.includes("\u65E0\u8BBF\u95EE\u6743\u9650")) {
          _flowusshared.out.err("\u6682\u65E0\u6743\u9650", `\u8BF7\u68C0\u67E5\u9875\u9762\u662F\u5426\u5F00\u542F\u516C\u5F00\u5206\u4EAB`);
        } else {
          _flowusshared.out.err("\u8BF7\u6C42\u9519\u8BEF", res.data.msg);
        }
        _process2.default.exit();
      }
      return res.data.data;
    });
  }
  getOssUrl(blocks) {
    return __async(this, null, function* () {
      const mediaBlocks = [];
      Object.keys(blocks).forEach((blockId) => {
        const block = blocks[blockId];
        if (block.type === 14) {
          mediaBlocks.push({
            blockId: block.uuid,
            ossName: block.data.ossName
          });
        }
        if (block.type === 0 && block.data.cover) {
          mediaBlocks.push({
            blockId: block.uuid,
            ossName: block.data.cover
          });
        }
      });
      const data = {
        batch: mediaBlocks
      };
      let medias = yield this._fetch("file/create_urls", { data, method: "POST" });
      medias = medias.map((item, index) => {
        return __spreadProps(__spreadValues({}, item), {
          blockId: mediaBlocks[index].blockId
        });
      });
      medias.forEach((media) => {
        const blockId = media.blockId;
        blocks[blockId].data.fullLink = media.url;
      });
      return blocks;
    });
  }
};


exports.FlowUsClient = FlowUsClient;
